source/common/ssl/context_impl.cc

```diff
@@ -370,6 +381,169 @@ ServerContextImpl::ServerContextImpl(ContextManagerImpl& parent, Stats::Scope& s
   }
+
+  if (!config.sessionTicketKeys().empty()) {
+    session_ticket_keys_.resize(config.sessionTicketKeys().size());
+    for (unsigned i = 0; i < config.sessionTicketKeys().size(); ++i) {
+      // If this changes, need to figure out how to deal with key files
+      // that previously worked.  For now, just assert so we'll notice that
+      // it changed if it does.
+      static_assert(sizeof(SessionTicketKey) == 80, "Input is expected to be this size");
+
+      const std::vector<uint8_t>& src_key = config.sessionTicketKeys().at(i);
+      SessionTicketKey& dst_key = session_ticket_keys_.at(i);
+
+      if (src_key.size() != sizeof(SessionTicketKey)) {
+        throw EnvoyException(fmt::format("Incorrect TLS session ticket key length.  "
+                                         "Index {}, length {}, expected length {}.",
+                                         i, src_key.size(), sizeof(SessionTicketKey)));
+      }
+
+      std::copy_n(src_key.begin(), dst_key.name.size(), dst_key.name.begin());
+      size_t pos = dst_key.name.size();
+      std::copy_n(src_key.begin() + pos, dst_key.hmac_key.size(), dst_key.hmac_key.begin());
+      pos += dst_key.hmac_key.size();
+      std::copy_n(src_key.begin() + pos, dst_key.aes_key.size(), dst_key.aes_key.begin());
+      pos += dst_key.aes_key.size();
+      ASSERT(src_key.begin() + pos == src_key.end());
+    }
+    SSL_CTX_set_tlsext_ticket_key_cb(
+        ctx_.get(),
+        [](SSL* ssl, uint8_t* key_name, uint8_t* iv, EVP_CIPHER_CTX* ctx, HMAC_CTX* hmac_ctx,
+           int encrypt) -> int {
+          ContextImpl* context_impl = static_cast<ContextImpl*>(
+              SSL_CTX_get_ex_data(SSL_get_SSL_CTX(ssl), sslContextIndex()));
+          return dynamic_cast<ServerContextImpl*>(context_impl)
+              ->sessionTicketProcess(ssl, key_name, iv, ctx, hmac_ctx, encrypt);
+        });
+  }
+
+  uint8_t session_context_buf[EVP_MAX_MD_SIZE] = {};
+  unsigned session_context_len = 0;
+  EVP_MD_CTX md;
+  int rc = EVP_DigestInit(&md, EVP_sha256());
+  RELEASE_ASSERT(rc == 1);
+
+  // Hash the CommonName/SANs in the server certificate.  This makes sure that
+  // sessions can only be resumed to a certificate for the same name, but allows
+  // resuming to unique certs in the case that different Envoy instances each have
+  // their own certs.
+  X509* cert = SSL_CTX_get0_certificate(ctx_.get());
+  X509_NAME* cert_subject = X509_get_subject_name(cert);
+  int cn_index = X509_NAME_get_index_by_NID(cert_subject, NID_commonName, -1);
+  X509_NAME_ENTRY* cn_entry = X509_NAME_get_entry(cert_subject, cn_index);
+  ASN1_STRING* cn_asn1 = X509_NAME_ENTRY_get_data(cn_entry);
+  RELEASE_ASSERT(ASN1_STRING_length(cn_asn1) > 0);
+  rc = EVP_DigestUpdate(&md, ASN1_STRING_data(cn_asn1), ASN1_STRING_length(cn_asn1));
+  RELEASE_ASSERT(rc == 1);
+
+  STACK_OF(GENERAL_NAME)* san_names = static_cast<STACK_OF(GENERAL_NAME)*>(
+      X509_get_ext_d2i(cert, NID_subject_alt_name, nullptr, nullptr));
```

Comment:
This should be:

    GENERAL_NAMES* san_names =
        static_cast<GENERAL_NAMES*>(X509_get_ext_d2i(cert, NID_subject_alt_name, nullptr, nullptr));

Reply:
Oh, that's much nicer.  Thanks!